### DISCUSSION TITLE
DeflatePerMessage VS memory consumption

### DISCUSSION THREAD

**Discussion by M1ha-Shvn** [Kludex/uvicorn#1850] [Potential Issue] [answered]
Hi.
I'm devloping a server serving websockets connections using FastAPI.
I've noticed, that creating several thousands of simultinious websocket connections leads to high memory usage (4 Gb per 5-8k websocket connections in my case). I've started debugging it with [tracemalloc](https://docs.python.org/3/library/tracemalloc.html) and found out that the largest amount of this memory is consumed by websockets deflate extension in [this line](https://github.com/aaugustin/websockets/blob/main/src/websockets/extensions/permessage_deflate.py#L64).
After that I've digged into [websockets deflate mechanics](https://websockets.readthedocs.io/en/stable/topics/compression.html) and found out that it can be tuned wisely in order to achive lower memory consumption using custom `ServerPerMessageDeflateFactory`. I've tried searching for it in FastApi => Starlette => Uvicorn code and it lead me [here](https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/websockets_impl.py#L98).
What is the source of memory leak:
1. Uvicorn creates separate `PerMessageDeflate` instance for each websocket connection (using `ServerPerMessageDeflateFactory`). From my point of view, it is a disadvantageous behaviour: it would be much better if it is created not for each websocket connection, but for each combination of connection parameters (like [Singleton](https://betterprogramming.pub/what-is-a-singleton-2dc38ca08e92?gi=1632bfde6b86) pattern, but created for each combination of parameters. Something like [lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache)).
2. The only setting uvicorn gives to tune deflate is just disabling it with `--ws-per-message-deflate`. It is not flexible for different cases.

**Comment by M1ha-Shvn** [comment, marked as answer]
Upd: disabling deflate solves leaking memory. So my hypothesis looks correct. Should I open an issue?

**Reply by Kludex** [reply]
Created https://github.com/encode/uvicorn/issues/1862 from this discussion.
